Small improvements in catalyst cli docs, option help, and intermediate file naming

mehrdad2m opened the pull request:
**Context:**
This PR fixes some small bugs or lack of documentation for catalyst-cli gathered from @dime10's feedbacks.

- The `--help` flag seems to dump a lot of unrelated options which makes it difficult to navigate through catalyst options.

- The possible stages for the --checkpoint-stage option are not mentioned in the documentation

- When using `--checkpoint-stage`, `--save-ir-after-each=pipeline` no longer works.

- The output from --save-ir-after-each=pass produces one output for each function when dealing with a function pass which results in a large number of outputs and one has to find the function of interest randomly.

**Description of the Change:**

- `--help` option now prints all the catalyst-cli specific options first before jumping into mlir-opt options

- Added more details to documentation

- Fixed the bug for `save-ir-after-each` and `save-ir-after-each` coexisting together.

- Output from `save-ir-after-each` now appends the name of the function to the file name making it easier to identify the desired output.

**Benefits:**
easier experience for catalyst-cli user

**Possible Drawbacks:**

**Related GitHub Issues:**

dime10 commented on mlir/lib/Driver/CompilerDriver.cpp:
```diff
@@ -861,26 +865,30 @@ int QuantumDriverMainFromCL(int argc, char **argv)
     // Any modifications made to the command-line interface should be documented in
     // doc/catalyst-cli/catalyst-cli.rst
-    cl::opt<std::string> WorkspaceDir("workspace", cl::desc("Workspace directory"), cl::init("."));
+    cl::OptionCategory CatalystCat("Catalyst-cli Options", "Frist print catalyst-cli options");
```
Don't think we need this string after looking at the output
```suggestion
    cl::OptionCategory CatalystCat("Catalyst-cli Options", "");
```

dime10 commented on mlir/lib/Driver/CompilerDriver.cpp:
```diff
@@ -904,8 +913,15 @@ int QuantumDriverMainFromCL(int argc, char **argv)
 
     // Register and parse command line options.
     std::string inputFilename, outputFilename;
+    std::string helpStr = "Catalyst Command Line Interface options. \n"
```
```suggestion
    std::string helpStr = "Catalyst Command Line Interface options.\n\n"
```

dime10 commented on mlir/lib/Driver/CompilerDriver.cpp:
```diff
@@ -904,8 +913,15 @@ int QuantumDriverMainFromCL(int argc, char **argv)
     // Register and parse command line options.
     std::string inputFilename, outputFilename;
+    std::string helpStr = "Catalyst Command Line Interface options. \n"
+                          "Below, there is a complete list of options for the Catalyst CLI tool\n"
+                          "In the first section, you can find the options that are used to\n"
+                          "configure the Catalyst compiler. Next, you can find the options\n"
+                          "specific to mlir-opt tool. Please note that we do not support all the\n"
+                          "options of mlir-opt tool. For the relavent options that are supported\n"
+                          "please refer to the documentation of the Catalyst CLI tool.\n";
```
Not sure what's common, should these be line wrapped by the terminal or by us? The paragraph just looks a little weird against the next one in the terminal:
```
Below, there is a complete list of options for the Catalyst CLI tool
In the first section, you can find the options that are used to
configure the Catalyst compiler. Next, you can find the options
specific to mlir-opt tool. Please note that we do not support all the
options of mlir-opt tool. For the relavent options that are supported
please refer to the documentation of the Catalyst CLI tool.

Available Dialects: acc, affine, amdgpu, amx, arith, arm_neon, arm_sme, arm_sve, async, bufferization, builtin, catalyst, cf, chlo, complex, dlti, emitc, func, gpu, gradient, index, irdl, linalg, llvm, math, memref, mesh, mhlo, mitigation, ml_program, mpi, nvgpu, nvvm, omp, pdl, pdl_interp, polynomial, quant, quantum, rocdl, scf, shape, sparse_tensor, spirv, stablehlo, tensor, tosa, transform, ub, vector, vhlo, x86vector, xegpu
USAGE: catalyst-cli [options] <input file>
```

mehrdad2m replied:
not too sure, I think it make sense to have some sort of line wrapping for the paragraph to be readable. the second paragraph is added inside an mlir-opt method so can't really fix the line wrapping there.

dime10 replied:
Well the line wrapping would otherwise be handled by the terminal to fit your window width.

mehrdad2m replied:
yeah that is true. I removed the wrapping to be consistent.

dime10 commented on mlir/lib/Driver/CompilerDriver.cpp:
```diff
@@ -904,8 +913,15 @@ int QuantumDriverMainFromCL(int argc, char **argv)
+                          "Below, there is a complete list of options for the Catalyst CLI tool\n"
+                          "In the first section, you can find the options that are used to\n"
+                          "configure the Catalyst compiler. Next, you can find the options\n"
+                          "specific to mlir-opt tool. Please note that we do not support all the\n"
+                          "options of mlir-opt tool. For the relavent options that are supported\n"
+                          "please refer to the documentation of the Catalyst CLI tool.\n";
```
What is meant here, which mlir-opt options don't we support?

mehrdad2m replied:
Right now, we are not exactly invoking the main function of mlir-opt, but we are its methods including the one that registers all of the options. so the functionality of the --tool=opt may not exactly be the same as mlir-opt in terms of options. One option that I found no working right now, and maybe worth adding in future is --split-input-file.

dime10 replied:
I see. The problem I found was that the CLI documentation does not talk about which options are and aren't supported. If it's rare that an option is unsupported I would consider removing that comment, or otherwise adding what we know about the support to the docs.

mehrdad2m replied:
the problem is that confirming exactly what option is working fine or not is not feasible since they are a huge number of them. Here I just want to warn that other options are not tested. I agree that we should talk more about the relevant mlir-opt options in the documentation. I will try to do that in another PR, and maybe clarify this there. in the meanwhile I'll remove the comment.

dime10 replied:
I think it's fine not to talk about it, I just found the comment confusing. Either way this is good as is 👍

dime10 reviewed (approved):
Nice work, these are great improvements 🚀 🚀 

I'm happy to merge as is, although for the help output I still wonder if we couldn't put the `mlir-opt` output behind an additional option, like `--help-passes`, `--help-opt`, `--help-tool` or something like that. The main `--help` option can then point out that there are additional help pages. It's just such a slog to scroll through all the output.

mehrdad2m commented on doc/catalyst-cli/catalyst-cli.rst:
```diff
@@ -138,7 +153,11 @@ Enable asynchronous QNodes.
 Define a *checkpoint stage*, used to indicate that the compiler should start only after reaching the
-given pass.
+given stage. The stages that are currently available are:
+* MLIR:: ``mlir``, and the name of piplines defined using the ``--catalyst-pipeline`` option.
```
hmm actually, I think maybe "mlir" and stage does not make sense for checkpoint stage. checkpoint stage means that we want to **start from** a certain stage and "mlir" is basically the first stage so it does not have any meaningful effect (It does not mean that we are doing all the mlir stages, for that we should use `--tool opt`)

dime10 replied:
I think the checkpoint stage is that we start _after_ the given stage no?

mehrdad2m replied:
yes that's correct. It is a bit redundant imo, but I fine with leaving it as is. However the change you suggested implies that using "mlir" only runs all the mlir stages which is misleading cause the llvm stages are still run

dime10 replied:
Right, maybe we could adjust the wording then, so it shouldn't be `(all MLIR stages)` but `(start with first MLIR stage)`?

joeycarter commented:
@mehrdad2m and @dime10, do we want to get this in for the release?

joeycarter commented on mlir/lib/Driver/CompilerDriver.cpp:
```diff
@@ -904,8 +913,13 @@ int QuantumDriverMainFromCL(int argc, char **argv)
+                          "Below, there is a complete list of options for the Catalyst CLI tool"
+                          "In the first section, you can find the options that are used to"
+                          "configure the Catalyst compiler. Next, you can find the options"
+                          "specific to mlir-opt tool.\n";
```
```suggestion
                          "specific to the mlir-opt tool.\n";
```

joeycarter reviewed (commented):
Thanks @mehrdad2m! Just had a few editorial suggestions.

mehrdad2m commented:
> @mehrdad2m and @dime10, do we want to get this in for the release?

Hi @joeycarter Yes I think it should be in the release. On my side it is ready to be merged. I can merged it after addressing your comments

joeycarter commented:
> Hi @joeycarter Yes I think it should be in the release. On my side it is ready to be merged. I can merged it after addressing your comments

Great, thanks! In that case, can you move the changelog entry to `doc/releases/changelog-0.10.0.md`?
